### PR TITLE
feat!: search the database by session name

### PR DESCRIPTION
**Pull request by dhess**
Apologies for the large first commit. There was no reasonable way to break it up into smaller pieces. Most of it is tests and refactoring, at least.

*edit*: note to future readers, we ended up making this implementation case-*in*sensitive only, at least for now. When time allows, we'll implement case-sensitive search. See https://github.com/hackworthltd/primer/pull/1033#issuecomment-1562089557 for the motivation.

### CODE REVIEW

**Comment by dhess** [comment]
Ready for review. I will add a matching case-insensitive search after I've tested out the frontend integration with this PR, but I don't expect these first 2 commits to change much.

**Comment by dhess** [comment]
Back to WIP status while I fix this up.

**Comment by dhess** [comment]
OK, this is proven to be working, and feasibly implemented, in the frontend now, so it's ready for review. I will add case-insensitive search next, though supporting both case-sensitive and -insensitive may prove to be a bit tricky on SQLite.

File: primer-rel8/src/Primer/Database/Rel8/Rel8Db.hs [hunk 1]
```diff
@@ -139,6 +141,26 @@ runRel8DbT m = runReaderT (unRel8DbT m)
...
+-- because a session name was invalid, so we opt for "convert it to a
+-- valid 'SessionName'".
+safeMkSession :: (SessionId, Text, UTCTime) -> Session
+safeMkSession (s, n, t) = Session s (safeMkSessionName n) (LastModified t)
+
+--
+-- It might be helpful if this function returned an indication of
+-- whether the original name was safe, but for now, we convert
+-- silently.
+
```
Comment (brprice) [inline primer-rel8/src/Primer/Database/Rel8/Rel8Db.hs:163]: This comment seems to have become detached from its code

File: primer-rel8/src/Primer/Database/Rel8/Rel8Db.hs [hunk 2]
```diff
@@ -395,6 +432,20 @@ sessionById :: UUID -> Query (Schema.SessionRow Expr)
...
+sessionByNameSubstr :: Text -> Query (Schema.SessionRow Expr)
+sessionByNameSubstr substr =
+  -- N.B. the arguments to 'like' are reversed from what you might expect.
+  filter (like (litExpr ("%" <> escape substr <> "%")) . Schema.name) =<< allSessions
+  where
+    -- Escape @%@ and @_@ characters in the given string, as these are
+    -- wildcards in SQL. Note that the backslash is the default escape
+    -- character, and therefore must also be escaped. Make sure we do
+    -- backslash first, as otherwise we'll double-escape the other two
+    -- replacements!
```
Comment (brprice) [inline primer-rel8/src/Primer/Database/Rel8/Rel8Db.hs:445]: ~Have you tested this with both the postgresql and sqlite dbs?~ Oh, I see: this code is only for the postgresql db.

File: primer-selda/src/Primer/Database/Selda/SQLite.hs [hunk 3]
```diff
@@ -200,34 +240,40 @@ instance MonadSeldaSQLiteDb m l => MonadDb (SeldaSQLiteDbT m) where
...
+    --
+    -- https://github.com/hackworthltd/primer/issues/1037
+    n' <- query $
+      Selda.aggregate $ do
+        session <- sessionByNameSubstr substr
+        pure $ Selda.count $ session ! #uuid
+    n <- case n' of
+      [n''] -> pure n''
+      -- something has gone terribly wrong: selda should never return
+      -- the empty find for a 'count' query.
```
Comment (brprice) [inline primer-selda/src/Primer/Database/Selda/SQLite.hs:268]: the empty /list/?

File: primer-rel8/test/Tests/FindSessions.hs [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    pAll <- findSessions substr $ OL{offset = 0, limit = Nothing}
+    total pAll @?= 0
+    pageContents pAll @?= expectedRows
+    step $ "Find all occurrences of " <> show substr <> " in session names (limit 50)"
+    p50 <- findSessions substr $ OL{offset = 0, limit = Just 50}
+    total p50 @?= 0
+    pageContents p50 @?= expectedRows
+
+test_findSessions_case_sensitive :: TestTree
+test_findSessions_case_sensitive = testCaseSteps "findSessions is case sensitive" $ \step' ->
```
[the remark this answers is not in the document]
Comment (dhess) [reply primer-rel8/test/Tests/FindSessions.hs:95]: Sure, but probably later. There are numerous corner cases to consider and so much time. Even the existing tests are hardly comprehensive. I believe the docs for `findSessions` mentions that case sensitivity is subject to implementation details and no promises are made wrt. that.

File: primer-selda/test/Tests/FindSessions.hs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    pAll <- findSessions substr $ OL{offset = 0, limit = Nothing}
+    total pAll @?= 0
+    pageContents pAll @?= expectedRows
+    step $ "Find all occurrences of " <> show substr <> " in session names (limit 50)"
+    p50 <- findSessions substr $ OL{offset = 0, limit = Just 50}
+    total p50 @?= 0
+    pageContents p50 @?= expectedRows
+
+test_findSessions_case_sensitive :: TestTree
+test_findSessions_case_sensitive = testCaseSteps "findSessions is case sensitive" $ \step' ->
```
Comment (brprice) [inline primer-selda/test/Tests/FindSessions.hs:95]: See comment about unicode & case sensitivity above

File: primer-selda/src/Primer/Database/Selda/SQLite.hs (hunk 3 above)
Comment (dhess) [reply primer-selda/src/Primer/Database/Selda/SQLite.hs:268]: Yes, thanks.

File: primer-service/src/Primer/Server.hs [hunk 6]
```diff
@@ -215,7 +216,7 @@ sessionsAPIServer :: ConvertServerLogs l => S.SessionsAPI (AsServerT (Primer l))
 sessionsAPIServer =
   S.SessionsAPI
     { S.createSession = newSession
-    , S.getSessionList = \p -> pagedDefault 100 p listSessions
+    , S.getSessionList = \s p -> pagedDefault 100 p $ maybe listSessions findSessions s
     , S.addSession = API.addSession
     , S.sessionAPI = sessionAPIServer
     }
```
Comment (brprice) [inline primer-service/src/Primer/Server.hs]: I'm confused here. The commit message says that in openapi we have one combined list+filter endpoint, but in "rich" api we have two separate ones. However, I thought this code was implementing the "rich" api, and it looks like we only have the combined one?

File: primer-rel8/test/Tests/FindSessions.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (brprice) [reply primer-rel8/test/Tests/FindSessions.hs:95]: > Sure, but probably later. There are numerous corner cases to consider and only so much time. Even the existing tests are hardly comprehensive. 

Fair enough

> I believe the docs for `findSessions` mention that case sensitivity for Unicode is subject to implementation details and no promises are made wrt. that.

Yes, they do. (I only saw this later in my review though)

File: primer-service/src/Primer/Server.hs (hunk 6 above)
Comment (dhess) [reply primer-service/src/Primer/Server.hs]: Rich meaning Haskell, not either of the HTTP APIs.

**Review by brprice — APPROVED** [review]
Approved modulo my minor comments above

File: primer-service/src/Primer/Server.hs (hunk 6 above)
Comment (brprice) [reply primer-service/src/Primer/Server.hs]: Oh, I see. Potentially worth rewording for clarity?

File: primer-rel8/src/Primer/Database/Rel8/Rel8Db.hs (hunk 1 above)
Comment (dhess) [reply primer-rel8/src/Primer/Database/Rel8/Rel8Db.hs:163]: Fixed.

File: primer-selda/src/Primer/Database/Selda/SQLite.hs (hunk 3 above)
Comment (dhess) [reply primer-selda/src/Primer/Database/Selda/SQLite.hs:268]: Fixed.

File: primer-rel8/test/Tests/FindSessions.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dhess) [reply primer-rel8/test/Tests/FindSessions.hs:95]: See https://github.com/hackworthltd/primer/issues/1039

**Comment by dhess** [comment]
In the interest of time pressure, I'm going to switch the implementation to be case-*in*sensitive, as I think that'll be much more frequently wanted by (and less surprising to) novice programmers. We can add a case-sensitive version later.
